File: java/503.next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        if (nums == null) {
            return null;
        }
        Deque<Integer> stack = new LinkedList<>();
        int[] ans = new int[nums.length];
        for (int i = nums.length * 2 - 1; i >= 0; i--) {
            int num = nums[i % nums.length];
            while (!stack.isEmpty() && stack.peekFirst() <= num) {
                stack.pollFirst();
            }
            ans[i % nums.length] = stack.isEmpty() ? -1 : stack.peekFirst();
            stack.offerFirst(num);
        }
        return ans;
    }
}

File: python3/503.next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [-1] * n
        
        nums = nums * 2
        
        stack = []
        
        for index, num in enumerate(nums):
            while stack and num > nums[stack[-1]]:
                last_index = stack.pop()
                ans[last_index] = num
                
            if index < n:
                stack.append(index)
                
        return ans
        
